src/craftdef.cpp

```diff
@@ -975,37 +975,20 @@ class CCraftDefManager: public IWritableCraftDefManager
 			IGameDef *gamedef) const
 	{
 		std::vector<CraftDefinition*> recipes_list;
-		CraftInput input;
-		CraftOutput tmpout;
-		tmpout.item = "";
-		tmpout.time = 0;
 
-		for(std::vector<CraftDefinition*>::const_reverse_iterator
-				i = m_craft_definitions.rbegin();
-				i != m_craft_definitions.rend(); i++)
-		{
-			CraftDefinition *def = *i;
+		std::map<std::string, std::vector<CraftDefinition*> >::const_iterator vecIter =
+			m_output_to_craft_definition_map.find(output.item);
 
-			/*infostream<<"Checking "<<input.dump()<<std::endl
-					<<" against "<<def->dump()<<std::endl;*/
+		if (vecIter == m_output_to_craft_definition_map.end())
+			return recipes_list;
 
-			try {
-				tmpout = def->getOutput(input, gamedef);
-				if(tmpout.item.substr(0,output.item.length()) == output.item)
-				{
-					// Get output, then decrement input (if requested)
-					input = def->getInput(output, gamedef);
-					recipes_list.push_back(*i);
-				}
-			}
-			catch(SerializationError &e)
-			{
-				errorstream<<"getCraftResult: ERROR: "
-						<<"Serialization error in recipe "
-						<<def->dump()<<std::endl;
-				// then go on with the next craft definition
-			}
+		for(std::vector<CraftDefinition*>::const_reverse_iterator
+			i = vecIter->second.rbegin();
+			i != vecIter->second.rend(); i++)
+		{
+			recipes_list.push_back(*i);
 		}
+			
```

Comment:
`get_all_craft_recipes` doesn't specify an order for the results, and `get_craft_recipe` can easily be changed to work with the opposite order if needed.  If the order is definitely important it might be better to store the recipes in reverse order so that in either case you can use a simple copy, or better yet return a const reference.  You should at least `.reserve()` the right amount of space if that won't work though.